### PR TITLE
[ZL-796] - Support Rest API migration

### PR DESCRIPTION
**Pull request by Varomir**
### Description
This PR changes how the JS-SDK handles API versions, to support easy migration from the legacy API Vapid to the new versioned API Rest. Read more about the Rest versioned API in our: [Confluence pages](https://procoretech.atlassian.net/wiki/spaces/DEV/pages/1360265934/Procore+Rest+API).

An optional `version` can now be passed from the library client-side or default `rest/v1.0` will be applied.

The API version can be specified at the time of the request.

| Example | Requested URL |
| --- | --- |
| `procore.get({base: '/me'})` | `https://app.procore.com/rest/v1.0/me` |
| `procore.get({base: '/me', version: 'v1.1'})` | `https://app.procore.com/rest/v1.1/me` |
| `procore.get({base: '/me', version: 'vapid'})` | `https://app.procore.com/vapid/me` |

### CODE REVIEW

File: lib/client.ts [hunk 1]
```diff
@@ -115,8 +115,19 @@ export class Client {
...
   )(this.host)
+
+  private version = (api_version: string): string => {
+    api_version = (api_version === 'undefined' || api_version === '') ? 'v1.0' : api_version;
+    if (api_version === 'vapid' || api_version === 'vapid/') {
+      return api_version
+    } else if(api_version.match(/v\d+\.\d+/)) {
+      return `rest/${api_version}`
+    } else {
+      throw new Error(`'${api_version}' is an invalid Procore API version`)
```
Comment (Varomir) [inline lib/client.ts:128]: As discussed, it would be nice to catch it o the `js-sdk-sample-app` side and show some alert/popup/banner message.

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@procore/js-sdk",
-  "version": "2.4.1",
+  "version": "2.5.0",
```
Comment (Varomir) [inline package.json:3]: So it should be version `3.0.0` right?

File: lib/client.ts [hunk 3]
```diff
@@ -115,8 +115,19 @@ export class Client {
       () => notNil(params.id),
       collectionUrl => `${collectionUrl}/${params.id}`
     ),
-    (hostname) => `${hostname}${S(base).template(params, '{', '}').s}`
+    (hostname) => `${hostname}/${this.version(api_version)}${S(base).template(params, '{', '}').s}`
   )(this.host)
+
+  private version = (api_version: string): string => {
```
Comment (Varomir) [inline lib/client.ts:121]: It would be better to refactor it with `Ramda` FP style. I just little bit stuck with that.

File: lib/client.ts [hunk 4]
```diff
@@ -102,7 +102,7 @@ export class Client {
     this.urlConfig
   )(endpoint)
 
-  private urlConfig = ({ base, action, params = {}, qs }: EndpointConfig): string => compose(
+  private urlConfig = ({ base, action, params = {}, qs, api_version }: EndpointConfig): string => compose(
```
Comment (zoolamyke) [inline lib/client.ts:105]: We should be using camelCase, I think

File: lib/client.ts [hunk 5]
```diff
@@ -115,8 +115,19 @@ export class Client {
...
     ),
-    (hostname) => `${hostname}${S(base).template(params, '{', '}').s}`
+    (hostname) => `${hostname}/${this.version(api_version)}${S(base).template(params, '{', '}').s}`
   )(this.host)
+
+  private version = (api_version: string): string => {
+    api_version = (api_version === 'undefined' || api_version === '') ? 'v1.0' : api_version;
+    if (api_version === 'vapid' || api_version === 'vapid/') {
+      return api_version
+    } else if(api_version.match(/v\d+\.\d+/)) {
```
Comment (zoolamyke) [inline lib/client.ts:125]:
```suggestion
    } else if (api_version.match(/v\d+\.\d+/)) {
```

Comment (zoolamyke) [inline lib/client.ts:125]: Would it be okay to put `^` in the beginning or a regex?

File: lib/client.ts (hunk 3 above)
Comment (zoolamyke) [inline lib/client.ts:121]: Line 122 can probably be removed
```suggestion
  private version = (api_version: string = 'v1.0'): string => {
```

Unless there is a valid case when `api_version` can equal `''`, in which case the following should do the job:

```
const finalizedApiVersion = apiVersion || 'v1.0'
```

File: lib/client.ts (hunk 5 above)
Comment (zoolamyke) [inline lib/client.ts:125]: `else` is redundant because of the `return` statement on line 124

File: package.json (hunk 2 above)
Comment (njbbaer) [reply package.json:3]: Yup

File: lib/client.ts [hunk 6]
```diff
@@ -88,35 +76,49 @@ export class Client {
...
+    }
+
+    if (notNil(qs)) {
+      url = `${url}?${stringify(qs, { arrayFormat: 'brackets' })}`;
+    }
+
+    return url;
+  };
+
+  private version = (apiVersion: string = 'v1.0'): string => {
```
Comment (Varomir) [inline lib/client.ts:110]: Do we need to process `''` (passed empty string from the client code) as default behaviour (`v1.0`) or we should throw an error in this case? (as now)

File: lib/client.ts [hunk 7]
```diff
@@ -88,35 +76,49 @@ export class Client {
...
-      () => notNil(params.id),
-      collectionUrl => `${collectionUrl}/${params.id}`
-    ),
-    (hostname) => `${hostname}${S(base).template(params, '{', '}').s}`
-  )(this.host)
+    this.authorize(this.request(this.url(endpoint), { method: 'DELETE', body: JSON.stringify(payload) }, reqConfig))
+
+  private url = (endpoint: Endpoint): string =>
+    notNil(endpoint) && (endpoint.constructor === String || endpoint instanceof String) ?
+      `${this.host}${endpoint}` : // TODO: Do we want to allow this. Version will not be handled for this
```
Comment (Varomir) [inline lib/client.ts:89]: This is probably one of the main questions for code review.

My vision: JS-SDK **should not** support the `endpoint instanceof String` login anymore

File: lib/clientOptions.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+export interface ClientOptions {
+  apiHostname?: string;
+  defaultVersion?: string;
+}
+
+export const ClientOptionsDefaults: ClientOptions = {
+  apiHostname: "https://app.procore.com",
```
Comment (Varomir) [inline lib/clientOptions.ts:7]: It may produce the `CORS policy` error:

`Access to fetch at 'https://app.procore.com/rest/v1.0/me?' from origin 'http://localhost:4050' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Credentials' header in the response is '' which must be 'true' when the request's credentials mode is 'include'.`

File: lib/refresher.ts [hunk 9]
```diff
@@ -11,18 +11,16 @@ class Refresher {
   }
 
   public authorize = (request: Function): Promise<any> => {
-    const self = this
+    const self = this;
```
Comment (Varomir) [inline lib/refresher.ts:14]: @gilmcquillan I suppose the first line `import 'isomorphic-fetch'` is **redundant** here, am I right?

File: README.md [hunk 10]
```diff
@@ -1,5 +1,6 @@
 # Procore JS SDK
 
+TODO: Is this circleci link available to external users?
```
Comment (Varomir) [inline README.md:3]: Yes, it's available. 

Just open it in incognito mode:
`<p><a href="https://circleci.com/gh/procore/js-sdk" rel="nofollow">CircleCI</a></p>`

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,4 +1,60 @@
...
+
+```javascript
+const procore = client(authorizer);
+
+// Default version is v1.0
+const { body } = await procore.get({ base: '/me' });
+// app.procore.com/rest/v1.0/me
+
+// Override default version
+const { body } = await procore.get({ base: '/me', version: 'v1.0' });
```
Comment (drymar) [inline CHANGELOG.md:34]:
```suggestion
const { body } = await procore.get({ base: '/me', version: 'v1.1' });
```

**Review by drymar — APPROVED** [review]

File: CHANGELOG.md (hunk 11 above)
Comment (gilmcquillan) [reply CHANGELOG.md:34]: I changed this to v1.0 since we don't have a v1.1 version of /me. This endpoint is not the best example. Let me see if I can find a better example.

File: package.json [hunk 12]
```diff
@@ -1,46 +1,44 @@
 {
   "name": "@procore/js-sdk",
-  "version": "2.4.1",
+  "version": "3.0.0",
   "description": "A wrapper for the procore API",
-  "main": "lib",
+  "main": "dist",
   "homepage": "https://github.com/procore/js-sdk",
```
Comment (davismartin) [inline package.json:6]: I realize this is a js-sdk, but do we want to also export the types as well?

**Review by davismartin — APPROVED** [review]
Just some small nitpicky changes that could be made around documentation and types I don't think they should necessarily block this going out so approved.

One other thing is I wasn't able to get sandbox to work with the sample app, but I am pretty sure the issue was in that repo.

**Review by drymar — APPROVED** [review]

**Review by davismartin — APPROVED** [review]

**Comment by Varomir** [comment]
LGTM

**Comment by bpross** [comment]
@njbbaer @gilmcquillan updated to send the sdk version in the headers. Let me know what you think about the naming.

File: package.json [hunk 13]
```diff
@@ -22,6 +22,7 @@
   "author": "Procore Tech <insights@procore.com> (http://procore.com)",
   "license": "MIT",
   "dependencies": {
+    "@procore/js-sdk": "^2.4.1",
```
Comment (gilmcquillan) [inline package.json:25]: Why is this here? Why would it have a dependency on itself?

Comment (bpross) [reply package.json]: I have removed these changes in addition to `yarn.lock`. It seems that the tests need the sdk in `./dist` and to accomplish that I ran `yarn add @procore/js-sdk` which in turn adds this sdk as a dep of itself.

File: package.json [hunk 14]
```diff
@@ -12,35 +12,27 @@
...
     "test:watch": "mocha -r ts-node/register test/**/*.spec.ts -w --watch-extensions ts,tsx",
     "watch": "tsc --watch",
-    "compile": "tsc",
-    "docs": "typedoc --includeDeclarations --excludeExternals --readme README.md --target es5 --out docs dist/*.ts",
-    "endpoints": "node-procore-endpoints lib"
+    "compile": "tsc"
   },
   "author": "Procore Tech <insights@procore.com> (http://procore.com)",
-  "license": "ISC",
+  "license": "MIT",
```
[the remark this answers is not in the document]
Comment (bpross) [reply package.json:23]: Interesting. Taking this PR over, so a bit unsure about this change. @Varomir @gilmcquillan any reason that the license here changed?

[the remark this answers is not in the document]
Comment (bpross) [reply package.json:23]: Given that the rest of our open source libraries use MIT, I think we should use that here as well.

Ruby SDK, MIT: https://github.com/procore/ruby-sdk/blob/master/LICENSE.txt
Blueprinter, MIT: https://github.com/procore/blueprinter/blob/master/MIT-LICENSE
Sift, MIT: https://github.com/procore/sift/blob/master/MIT-LICENSE

I will open up a discussion with legal around what license we should be using.

File: CHANGELOG.md [hunk 15]
```diff
@@ -1,4 +1,77 @@
 # Change Log
 
+## 3.0.0 (March 31, 2021)
+
+* Adds support for Rest API versioning (rest/v1.0 is)
+* Remove `ramda` dependency
+* Remove `string` dependency
+* Remove `@procore/js-sdk-endpoints` dependency
+* Add support for revoking token `revoke()`
```
Comment (gilmcquillan) [inline CHANGELOG.md:9]: Add another bullet for the new header

Comment (gilmcquillan) [reply CHANGELOG.md:9]: Add

```
* Updated `destroy` method to `delete` method.
```

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,4 +1,77 @@
...
+
+### Breaking Changes
+
+* Remove `/vapid` from base attribute passed to `client.<method>` functions. Versioning will be handled in the `client.<method>` function. See the [Upgrading](#upgrading) section for details about how to upgrade to version 3.0.0.
+
```
Comment (gilmcquillan) [inline CHANGELOG.md:14]: Add

```
* `destroy` method is no longer supported. Convert all `destroy` requests to `delete`.
```

File: lib/client.ts [hunk 17]
```diff
@@ -31,41 +24,41 @@ interface SDKResponse {
...
 }
 
 const baseRequest = (defaults: RequestInit): Function => (url: string, config: RequestInit, reqConfig?: RequestConfig): Function => {
-  const headers = new Headers()
-  headers.append('Accept', 'application/json')
-  headers.append('Content-Type', 'application/json')
+  const headers = new Headers();
+  headers.append('Accept', 'application/json');
+  headers.append('Content-Type', 'application/json');
+  headers.append('X-Procore-Sdk-Version', sdkVersionHeader);
```
Comment (gilmcquillan) [inline lib/client.ts:42]: Let's update this to `Procore-Sdk-Version`. I've updated https://procoretech.atlassian.net/wiki/spaces/DEV/pages/1458995266/Procore%2BREST%2BAPI%2BRequirements#SHOULD-use-only-the-specified-proprietary-Procore-headers-Rest%2FV2%2B to reflect this change as well.

Comment (gilmcquillan) [reply lib/client.ts:42]: Update the readme as well.

**Comment by Varomir** [comment]
@bross I've got a compilation error after `yarn & yarn test`
```
TSError: ⨯ Unable to compile TypeScript:
test/authorize.spec.ts:2:27 - error TS2307: Cannot find module '../dist/index' or its corresponding type declarations.

2 import { authorize } from '../dist/index'
                            ~~~~~~~~~~~~~~~
```

**Comment by bpross** [comment]
> @bross I've got a compilation error after `yarn & yarn test`
> 
> ```
> TSError: ⨯ Unable to compile TypeScript:
> test/authorize.spec.ts:2:27 - error TS2307: Cannot find module '../dist/index' or its corresponding type declarations.
> 
> 2 import { authorize } from '../dist/index'
>                             ~~~~~~~~~~~~~~~
> ```
> 
> ```
> js-sdk % node --version
> v10.22.1
> js-sdk % npm --version 
> 6.14.6
> ```

Thanks for pointing this out @Varomir 

After some investigation, this appears to be a side effect of importing `package.json` to get the SDK version. We have two options here: we can either modify the tests to reference the new location `dist/lib/*` or create a new file that stores the SDK version and read from that. Will discuss in standup.

File: package.json [hunk 18]
```diff
@@ -1,46 +1,45 @@
 {
   "name": "@procore/js-sdk",
-  "version": "2.4.1",
+  "version": "3.0.0",
   "description": "A wrapper for the procore API",
-  "main": "lib",
+  "main": "dist/lib",
```
Comment (njbbaer) [inline package.json:5]: When does `dist` get built? Is it created automatically when the user installs/runs the package?

This article suggests making the files in `dist` available to NPM using the `files` key:
https://mykeels.medium.com/node-how-to-handle-your-dist-files-9667f47316ab

**Review by gilmcquillan — APPROVED** [review]
